File: consolidado/tools/ghidra_scripts/ExtractMtsNamespace.py
# Ghidra headless script: Extrai decompilação C de funções MTS/GBE/ICC/glue
# Uso:
#   analyzeHeadless <project_loc> <project> -process kmem_dump_1252.bin \
#       -postScript ExtractMtsNamespace.py \
#       -scriptPath /mnt/t/downloads/PS4/linux_in_ps4/consolidado/tools/ghidra_scripts \
#       -readOnly
#
# Saida: um arquivo .c por função, em consolidado/decompiled/extracted/

#@category PS4-MTS
#@author opencode-automation

import os
import re
from ghidra.app.decompiler import DecompInterface
from ghidra.util.task import ConsoleTaskMonitor

OUT_DIR = "/mnt/t/downloads/PS4/linux_in_ps4/consolidado/decompiled/extracted"
KNOWN_DIRS = [
    "/mnt/t/downloads/PS4/linux_in_ps4/consolidado/decompiled",
    "/mnt/t/downloads/PS4/linux_in_ps4/consolidado/decompiled/legacy_raiz",
]

# Ranges de endereços (low, high) cobrindo namespaces de interesse.
# Formato: (low64, high64) exclusive high.
TARGET_RANGES = [
    (0xffffffffdc478000, 0xffffffffdc479000),  # ICC alias
    (0xffffffffdc3f5000, 0xffffffffdc3f6000),  # icc_query wrapper
    (0xffffffffdc526000, 0xffffffffdc52b000),  # GBE clk/phy/domain
    (0xffffffffdc528000, 0xffffffffdc529000),  # icc_power dispatcher (inside dc52)
    (0xffffffffdc530000, 0xffffffffdc539000),  # GBE aux
    (0xffffffffdc536000, 0xffffffffdc538000),  # overlaps dc53 aux
    (0xffffffffdc5a0000, 0xffffffffdc5b0000),  # MTS driver completo (dc5axxxx)
    (0xffffffffdc5b0000, 0xffffffffdc5c0000),  # MTS helpers (dc5bxxxx)
    (0xffffffffdc6df000, 0xffffffffdc6e0000),  # glue block reset
    (0xffffffffdc718000, 0xffffffffdc71a000),  # glue write + baikal pcie + clock init
    (0xffffffffdc7c8000, 0xffffffffdc7c9000),  # ICC devpower
]

# Endereços-base já extraídos (preenchido por scan_known)
KNOWN_DONE = set()

ADDR_RE = re.compile(
    r"0xffffffff"
    r"(dc5[0-9a-f]{4}"
    r"|dc7c8[0-9a-f]{3}"
    r"|dc478[0-9a-f]{3}"
    r"|dc3f5[0-9a-f]{3}"
    r"|dc6df[0-9a-f]{3}"
    r"|dc718[0-9a-f]{3}"
    r"|dc719[0-9a-f]{3})"
)

def scan_known():
    """Coleta enderecos extraidos do header de cada .txt existente."""
    for d in KNOWN_DIRS:
        if not os.path.isdir(d):
            continue
        for fn in os.listdir(d):
            if not fn.endswith(".txt"):
                continue
            path = os.path.join(d, fn)
            try:
                with open(path, "r") as f:
                    head = f.read(2048)
                m = ADDR_RE.search(head)
                if m:
                    s = m.group(1)
                    KNOWN_DONE.add(int(s, 16))
            except Exception:
                pass

def in_targets(addr):
    a = addr & 0xffffffffffffffff
    for (lo, hi) in TARGET_RANGES:
        if lo <= a < hi:
            return True
    return False

def short_name(addr):
    """dc5a0070 style."""
    s = "%016x" % (addr & 0xffffffffffffffff)
    # s = 16 hex chars. Quero os últimos 8 (dc + 6 nibbles high precisam aparecer).
    # 0xffffffffdc5a0070 -> ultimos 8: dc5a0070
    return s[-8:]

def main():
    if not os.path.isdir(OUT_DIR):
        os.makedirs(OUT_DIR)
    scan_known()

    program = currentProgram
    fm = program.getFunctionManager()
    monitor = ConsoleTaskMonitor()

    decomp = DecompInterface()
    decomp.openProgram(program)

    funcs = []
    func_iter = fm.getFunctions(True)
    while func_iter.hasNext():
        f = func_iter.next()
        addr = f.getEntryPoint().getOffset()
        if in_targets(addr):
            funcs.append(f)

    print("=" * 60)
    print("ExtractMtsNamespace: %d funcoes no escopo alvo" % len(funcs))
    print("=" * 60)

    extracted = 0
    skipped = 0
    failed = 0

    for f in funcs:
        addr = f.getEntryPoint().getOffset()
        sym = short_name(addr)
        # Skip se ja extraido (compara low bits 20 ou 24)
        already = False
        for d in KNOWN_DIRS + [OUT_DIR]:
            cand = os.path.join(d, "decompiled_%s.txt" % sym)
            legacy_cand = os.path.join(d, "%s.txt" % sym)
            if os.path.exists(cand) or os.path.exists(legacy_cand):
                already = True
                break
        # Também trata caso do nome ja ter prefix
        for d in KNOWN_DIRS + [OUT_DIR]:
            for suffix in ["_%s" % sym, "dc%s" % sym[2:]]:
                pass
        if already:
            skipped += 1
            continue

        out_file = os.path.join(OUT_DIR, "decompiled_%s.txt" % sym)
        try:
            res = decomp.decompileFunction(f, 120, monitor)
            if res and res.decompileCompleted():
                src = res.getDecompiledFunction().getC()
                with open(out_file, "w") as fh:
                    fh.write("// Extraido por Ghidra headless (ExtractMtsNamespace.py)\n")
                    fh.write("// addr: 0x%016x  name: %s  size: %d\n" % (
                        addr, f.getName(), f.getBody().getNumAddresses()))
                    fh.write("// escopo: namespace MTS/GBE/ICC/glue\n")
                    fh.write("// chamadores: %d  chamadas: %d\n" % (
                        len(f.getCallingFunctions()), len(f.getCalledFunctions())))
                    fh.write("\n")
                    fh.write(src.encode('ascii', 'replace'))
                extracted += 1
                print("[OK] 0x%016x -> %s" % (addr, os.path.basename(out_file)))
            else:
                failed += 1
                print("[FAIL] 0x%016x decompile incomplete" % addr)
        except Exception as e:
            failed += 1
            print("[EXC] 0x%016x %s" % (addr, str(e)))

    decomp.disposeProgram(program)

    summary = os.path.join(OUT_DIR, "_extraction_summary.txt")
    with open(summary, "w") as f:
        f.write("Resumo ExtractMtsNamespace.py\n")
        f.write("=" * 40 + "\n")
        f.write("Total no escopo:        %d\n" % len(funcs))
        f.write("Extraidas agora:        %d\n" % extracted)
        f.write("Skipped (ja existiam):  %d\n" % skipped)
        f.write("Falhas de decomp:       %d\n" % failed)
    print("=" * 60)
    print("FINAL: extracted=%d  skipped=%d  failed=%d  total=%d" % (
        extracted, skipped, failed, len(funcs)))
    print("summary: %s" % summary)

main()

File: consolidado/tools/ghidra_scripts/ExtractMtsNamespace.java
// Ghidra headless script (Java) - extrai decompilação C de funções MTS/GBE/ICC/glue
// sem dependencia de Python/PyGhidra.
//
// Uso:
//   analyzeHeadless <project_loc> <project> -process kmem_dump_1252.bin \
//       -postScript ExtractMtsNamespace.java \
//       -scriptPath /mnt/t/downloads/PS4/linux_in_ps4/consolidado/tools/ghidra_scripts \
//       -readOnly -noanalysis
//
//@category PS4-MTS
//@author opencode-automation

import ghidra.app.script.GhidraScript;
import ghidra.app.decompiler.DecompInterface;
import ghidra.app.decompiler.DecompileOptions;
import ghidra.app.decompiler.DecompileResults;
import ghidra.app.decompiler.DecompiledFunction;
import ghidra.app.cmd.disassemble.DisassembleCommand;
import ghidra.app.cmd.function.CreateFunctionCmd;
import ghidra.program.model.listing.Function;
import ghidra.program.model.listing.FunctionManager;
import ghidra.program.model.listing.Listing;
import ghidra.program.model.address.Address;
import ghidra.program.model.address.AddressFactory;
import ghidra.program.model.address.AddressSet;
import ghidra.program.model.address.AddressRange;
import ghidra.util.task.ConsoleTaskMonitor;
import ghidra.util.task.TaskMonitor;

import java.io.File;
import java.io.FileWriter;
import java.io.PrintWriter;
import java.io.IOException;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

public class ExtractMtsNamespace extends GhidraScript {

    private static final String OUT_DIR = "/mnt/t/downloads/PS4/linux_in_ps4/consolidado/decompiled/extracted";

    // Endereços-alvo das LACUNAS (funções ainda não decompiladas, validadas em testes ao vivo).
    // Para adicionar nova função: inserir endereço aqui.
    private static final long[] TARGET_ADDRS = {
        // ===== MTS driver - lacunas validadas em test_history =====
        0xffffffffdc5a2840L, // MDIO read high word
        0xffffffffdc5a2950L, // MDIO write opcode 0x2000
        0xffffffffdc5a4950L, // gatilho BAR0+0x1c = 0x80000000
        0xffffffffdc5a4e90L, // relacionado RMU
        0xffffffffdc5a5050L, // provável pós-trigger
        0xffffffffdc5a5200L, // RMU sub-header 0x9807
        0xffffffffdc5a6290L, // sub-rotina
        // ===== MTS helpers =====
        0xffffffffdc5ba8d0L, // aloca BARs
        0xffffffffdc5baa30L, // cria ifnet
        // ===== Glue/PCIe sub-funções =====
        0xffffffffdc6dfb60L, // primitiva reset glue
        0xffffffffdc7187a0L, 0xffffffffdc7187d0L, 0xffffffffdc718800L, // glue read/write
        // ===== ICC =====
        0xffffffffdc3f5bd0L, // wrapper icc_query(4, 0x38) FUNDAMENTAL
        0xffffffffdc574150L, // registra handlers ICC
        0xffffffffdc528ef0L, // handler 4/0x38 GBE power-on
        // ===== GBE clk/phy =====
        0xffffffffdc529ed0L, 0xffffffffdc529f40L, 0xffffffffdc52a4f0L,
    };

    @Override
    public void run() throws Exception {
        File outDir = new File(OUT_DIR);
        if (!outDir.exists()) {
            outDir.mkdirs();
        }

        DecompInterface decomp = new DecompInterface();
        DecompileOptions opts = new DecompileOptions();
        opts.setDecompilerMaxTimeout(60);
        decomp.setOptions(opts);
        decomp.openProgram(currentProgram);

        FunctionManager fm = currentProgram.getFunctionManager();
        Listing listing = currentProgram.getListing();
        AddressFactory af = currentProgram.getAddressFactory();
        TaskMonitor monitor = getMonitor();

        int extracted = 0, skipped = 0, failed = 0, created = 0, notInMem = 0;

        println("========================================================");
        println("ExtractMtsNamespace (Java): " + TARGET_ADDRS.length + " enderecos alvo");
        println("========================================================");

        for (long addrInt : TARGET_ADDRS) {
            Address addr = af.getDefaultAddressSpace().getAddress(addrInt);

            // Verifica se endereço existe na memoria
            if (!currentProgram.getMemory().contains(addr)) {
                notInMem++;
                println("[NOT-IN-MEM] 0x" + Long.toHexString(addrInt));
                continue;
            }

            String shortName = String.format("%016x", addrInt).substring(8);

            // Tenta obter função existente
            Function f = fm.getFunctionAt(addr);
            if (f == null) {
                // Disassemble em torno do endereço (256 bytes)
                Address endAddr = af.getDefaultAddressSpace().getAddress(addrInt + 255);
                AddressSet set = new AddressSet(addr, endAddr);
                DisassembleCommand disasm = new DisassembleCommand(set, false);
                disasm.enableCodeAnalysis(false);
                disasm.applyTo(currentProgram, monitor);

                // Agora tenta criar função
                CreateFunctionCmd cmd = new CreateFunctionCmd(addr);
                if (cmd.applyTo(currentProgram, monitor)) {
                    created++;
                    f = fm.getFunctionAt(addr);
                } else {
                    failed++;
                    println("[CREATE-FAIL] 0x" + Long.toHexString(addrInt) + " " + cmd.getStatusMsg());
                    continue;
                }
            }

            if (f == null) {
                failed++;
                println("[NO-FUNC] 0x" + Long.toHexString(addrInt));
                continue;
            }

            String outFileName = "decompiled_" + shortName + ".txt";
            File outFile = new File(outDir, outFileName);
            if (outFile.exists()) {
                skipped++;
                continue;
            }

            try {
                DecompileResults res = decomp.decompileFunction(f, 90, monitor);
                if (res != null && res.decompileCompleted()) {
                    DecompiledFunction df = res.getDecompiledFunction();
                    String src = df.getC();

                    PrintWriter pw = new PrintWriter(new FileWriter(outFile));
                    pw.println("// Extraido por Ghidra headless (ExtractMtsNamespace.java)");
                    pw.println("// addr: 0x" + Long.toHexString(addrInt) + "  name: " + f.getName()
                            + "  size: " + f.getBody().getNumAddresses());
                    pw.println("// escopo: lacuna MTS/GBE/ICC/glue identificada em testes ao vivo");
                    pw.println("// chamadores (" + f.getCallingFunctions().size() + "):");
                    int cc = 0;
                    for (Function c : f.getCallingFunctions()) {
                        if (cc < 20) {
                            pw.println("//   " + c.getName() + " @ 0x" + Long.toHexString(c.getEntryPoint().getOffset()));
                        }
                        cc++;
                    }
                    if (cc > 20) pw.println("//   ... +" + (cc - 20) + " mais");
                    pw.println("// chamadas (" + f.getCalledFunctions().size() + "):");
                    int cd = 0;
                    for (Function c : f.getCalledFunctions()) {
                        if (cd < 20) {
                            pw.println("//   " + c.getName() + " @ 0x" + Long.toHexString(c.getEntryPoint().getOffset()));
                        }
                        cd++;
                    }
                    if (cd > 20) pw.println("//   ... +" + (cd - 20) + " mais");
                    pw.println();
                    pw.println(src);
                    pw.close();

                    extracted++;
                    println("[OK] 0x" + Long.toHexString(addrInt) + " -> " + outFileName
                            + "  callers=" + f.getCallingFunctions().size()
                            + "  callees=" + f.getCalledFunctions().size());
                } else {
                    failed++;
                    println("[DECOMP-FAIL] 0x" + Long.toHexString(addrInt));
                }
            } catch (Exception e) {
                failed++;
                println("[EXC] 0x" + Long.toHexString(addrInt) + " " + e.getMessage());
            }
        }

        decomp.disposeProgram(currentProgram);

        File summary = new File(outDir, "_extraction_summary.txt");
        try (PrintWriter s = new PrintWriter(new FileWriter(summary))) {
            s.println("Resumo ExtractMtsNamespace.java");
            s.println("==================================================");
            s.println("Enderecos-alvo:           " + TARGET_ADDRS.length);
            s.println("Extraidos agora:          " + extracted);
            s.println("Skipped (ja existiam):    " + skipped);
            s.println("Falhas (decomp):          " + failed);
            s.println("Fora da_MEM do program:   " + notInMem);
            s.println("Funcoes criadas (CreateFunctionCmd): " + created);
        }
        println("========================================================");
        println("FINAL: extracted=" + extracted + "  skipped=" + skipped
                + "  failed=" + failed + "  created=" + created + "  notInMem=" + notInMem);
    }
}
